source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+
+void HappyEyeballsConnectionImpl::connect() {
+  ENVOY_BUG(!connect_finished_, "connection already connected");
+  connections_[0]->connect();
+  maybeScheduleNextAttempt();
+}
+
+void HappyEyeballsConnectionImpl::addWriteFilter(WriteFilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->addWriteFilter(filter);
+    return;
+  }
+  post_connect_state_.write_filters_.push_back(filter);
+}
+
+void HappyEyeballsConnectionImpl::addFilter(FilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->addFilter(filter);
+    return;
+  }
+  post_connect_state_.filters_.push_back(filter);
+}
+
+void HappyEyeballsConnectionImpl::addReadFilter(ReadFilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->addReadFilter(filter);
+    return;
+  }
+  post_connect_state_.read_filters_.push_back(filter);
+}
+
+void HappyEyeballsConnectionImpl::removeReadFilter(ReadFilterSharedPtr filter) {
+  if (connect_finished_) {
+    connections_[0]->removeReadFilter(filter);
+    return;
+  }
+  auto i = post_connect_state_.read_filters_.begin();
+  while (i != post_connect_state_.read_filters_.end()) {
+    if (*i == filter) {
+      post_connect_state_.read_filters_.erase(i);
+      return;
+    }
+  }
+  ASSERT(false);
+}
+
+bool HappyEyeballsConnectionImpl::initializeReadFilters() {
+  if (connect_finished_) {
+    return connections_[0]->initializeReadFilters();
+  }
+  if (!post_connect_state_.read_filters_.empty()) {
+    return false;
+  }
+  post_connect_state_.initialize_read_filters_.value() = true;
+  return true;
+}
+
+void HappyEyeballsConnectionImpl::addBytesSentCallback(Connection::BytesSentCb cb) {
+  if (connect_finished_) {
+    connections_[0]->addBytesSentCallback(cb);
+    return;
+  }
+  post_connect_state_.bytes_sent_callbacks_.push_back(cb);
+}
+
+void HappyEyeballsConnectionImpl::enableHalfClose(bool enabled) {
+  if (!connect_finished_) {
+    per_connection_state_.enable_half_close_ = enabled;
+  }
+  for (auto& connection : connections_) {
+    connection->enableHalfClose(enabled);
+  }
+}
+
+bool HappyEyeballsConnectionImpl::isHalfCloseEnabled() {
+  if (connect_finished_) {
+    return connections_[0]->isHalfCloseEnabled();
+  }
+  return per_connection_state_.enable_half_close_.has_value() &&
+         per_connection_state_.enable_half_close_.value();
+}
+
+std::string HappyEyeballsConnectionImpl::nextProtocol() const {
+  if (connect_finished_) {
+    return connections_[0]->nextProtocol();
+  }
+  return "";
+}
+
+void HappyEyeballsConnectionImpl::noDelay(bool enable) {
+  if (!connect_finished_) {
+    per_connection_state_.no_delay_ = enable;
+  }
+  for (auto& connection : connections_) {
+    connection->noDelay(enable);
+  }
+}
+
+void HappyEyeballsConnectionImpl::readDisable(bool disable) {
+  if (connect_finished_) {
+    connections_[0]->readDisable(disable);
+    return;
+  }
+  if (!post_connect_state_.read_disable_count_.has_value()) {
```

Comment:
optional: I know you'll have to do the math here regardless (for new connections) but have you considered passing it on live to connections? I don't know why anyone would read disable while connecting but if they did arguably delaying passing it on could be problematic?

Reply:
I'm not sure if that'll work?  I think for TLS the Connected event is raised when handshake is complete and stopping reads is ill advised.
honestly I think no one calls readDisable before connected so it doesn't matter either way